<?xml version="1.0" ?>
<robot name="panda">
  <link name="world">
    <inertial>
      <mass value="1"/>
      <inertia ixx="0.04" ixy="0.0" ixz="0.0" iyy="0.04" iyz="0.0" izz="0.2"/>
    </inertial>
  </link>
  <link name="table">
    <visual>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/table/visual/table.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/table/collision/table.obj"/>
      </geometry>
    </collision>
  </link>
  <joint name="world_joint" type="fixed">
    <parent link="world"/>
    <child link="panda_link0"/>
    <origin xyz="-0.43 0.0 0.01" rpy="0.0 0.0 0.0"/>
  </joint>
  <joint name="table_joint" type="fixed">
    <parent link="world"/>
    <child link="table"/>
    <origin xyz="0.0 0.0 -0.39" rpy="0.0 0.0 0.0"/>
  </joint>

  <link name="panda_link0">
    <visual>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/link0.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/collision/link0.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="panda_link1">
    <visual>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/link1.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/collision/link1.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint1" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
    <origin rpy="0 0 0" xyz="0 0 0.333"/>
    <parent link="panda_link0"/>
    <child link="panda_link1"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" lower="-2.8973" upper="2.8973" velocity="2.1750"/>
  </joint>
  <link name="panda_link2">
    <visual>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/link2.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/collision/link2.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint2" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-1.7628" soft_upper_limit="1.7628"/>
    <origin rpy="-1.5707963267948966 0 0" xyz="0 0 0"/>
    <parent link="panda_link1"/>
    <child link="panda_link2"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" lower="-1.7628" upper="1.7628" velocity="2.1750"/>
  </joint>
  <link name="panda_link3">
    <visual>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/link3.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/collision/link3.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint3" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0 -0.316 0"/>
    <parent link="panda_link2"/>
    <child link="panda_link3"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" lower="-2.8973" upper="2.8973" velocity="2.1750"/>
  </joint>
  <link name="panda_link4">
    <visual>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/link4.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/collision/link4.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint4" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-3.0718" soft_upper_limit="-0.0698"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0.0825 0 0"/>
    <parent link="panda_link3"/>
    <child link="panda_link4"/>
    <axis xyz="0 0 1"/>
    <limit effort="87" lower="-3.0718" upper="-0.0698" velocity="2.1750"/>
  </joint>
  <link name="panda_link5">
    <visual>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/link5.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/collision/link5.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint5" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
    <origin rpy="-1.5707963267948966 0 0" xyz="-0.0825 0.384 0"/>
    <parent link="panda_link4"/>
    <child link="panda_link5"/>
    <axis xyz="0 0 1"/>
    <limit effort="12" lower="-2.8973" upper="2.8973" velocity="2.6100"/>
  </joint>
  <link name="panda_link6">
    <visual>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/link6.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/collision/link6.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint6" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-0.0175" soft_upper_limit="3.7525"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0 0 0"/>
    <parent link="panda_link5"/>
    <child link="panda_link6"/>
    <axis xyz="0 0 1"/>
    <limit effort="12" lower="-0.0175" upper="3.7525" velocity="2.6100"/>
  </joint>
  <link name="panda_link7">
    <visual>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/link7.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/collision/link7.stl"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_joint7" type="revolute">
    <safety_controller k_position="100.0" k_velocity="40.0" soft_lower_limit="-2.8973" soft_upper_limit="2.8973"/>
    <origin rpy="1.5707963267948966 0 0" xyz="0.088 0 0"/>
    <parent link="panda_link6"/>
    <child link="panda_link7"/>
    <axis xyz="0 0 1"/>
    <limit effort="12" lower="-2.8973" upper="2.8973" velocity="2.6100"/>
  </joint>
  <link name="panda_link8"/>
  <joint name="panda_joint8" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.107"/>
    <parent link="panda_link7"/>
    <child link="panda_link8"/>
    <axis xyz="0 0 0"/>
  </joint>
  <joint name="panda_hand_joint" type="fixed">
    <parent link="panda_link8"/>
    <child link="panda_hand"/>
    <origin rpy="0 0 -0.7853981633974483" xyz="0 0 0"/>
  </joint>
  <link name="panda_hand">
    <visual>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/hand.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/collision/hand.stl"/>
      </geometry>
    </collision>
  </link>
  <link name="panda_leftfinger">
    <visual>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/finger.dae"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/finger.dae"/>
      </geometry>
    </collision>
  </link>
  <link name="panda_rightfinger">
    <visual>
      <origin rpy="0 0 3.141592653589793" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/finger.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 3.141592653589793" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://sapien_simulator/meshes/visual/finger.dae"/>
      </geometry>
    </collision>
  </link>
  <joint name="panda_finger_joint1" type="prismatic">
    <parent link="panda_hand"/>
    <child link="panda_leftfinger"/>
    <origin rpy="0 0 0" xyz="0 0 0.0584"/>
    <axis xyz="0 1 0"/>
    <limit effort="20" lower="0.0" upper="0.04" velocity="0.2"/>
  </joint>
  <joint name="panda_finger_joint2" type="prismatic">
    <parent link="panda_hand"/>
    <child link="panda_rightfinger"/>
    <origin rpy="0 0 0" xyz="0 0 0.0584"/>
    <axis xyz="0 -1 0"/>
    <limit effort="20" lower="0.0" upper="0.04" velocity="0.2"/>
    <mimic joint="panda_finger_joint1"/>
  </joint>
  <link name="panda_ee_link"></link>
  <joint name="panda_ee_link_joint" type="fixed">
    <origin rpy="0.0 0.0 -0.785" xyz="0.0 0.0 0.15"/>
    <parent link="panda_link8"/>
    <child link="panda_ee_link"/>
  </joint>

  <!--  kinect -->
  <material name="black">
    <color rgba="0. 0. 0. 1"/>
  </material>
  <material name="dark_grey">
    <color rgba="0.1 0.1 0.1 1"/>
  </material>
  <material name="grey">
    <color rgba="0.6 0.6 0.6 1"/>
  </material>
  <material name="white">
    <color rgba="0.9 0.9 0.9 1"/>
  </material>
  <link name="kinect_body">
    <visual>
      <origin xyz="0. 0. 0."/>
      <geometry>
        <box size="0.0994 0.0996 0.023"/>
      </geometry>
      <material name="white"/>
    </visual>
  </link>
  <link name="kinect_base">
    <visual>
      <origin xyz="-0.013 0. 0."/>
      <geometry>
        <box size="0.026 0.103 0.039"/>
      </geometry>
      <material name="grey"/>
    </visual>
  </link>
  <link name="kinect_visor">
    <visual>
      <origin xyz="-0.0128 0. 0."/>
      <geometry>
        <box size="0.026 0.101 0.037"/>
      </geometry>
      <material name="black"/>
    </visual>
  </link>
  <joint name="kinect_base_to_body" type="fixed">
    <parent link="kinect_base"/>
    <child link="kinect_body"/>
    <origin rpy="0. 0. 0." xyz="-0.0757 0. 0.008"/>
  </joint>
  <joint name="kinect_base_to_visor" type="fixed">
    <parent link="kinect_base"/>
    <child link="kinect_visor"/>
    <origin rpy="0. 0. 0." xyz="0. 0. 0."/>
  </joint>
  <joint name="kinect__visor_to_depth" type="fixed">
    <parent link="kinect_visor"/>
    <child link="depth_camera_link"/>
    <origin rpy="-1.57 0. -1.57" xyz="0. 0. 0."/>
  </joint>
  <link name="rgb_camera_link"/>
  <link name="depth_camera_link"/>
  <joint name="depth_to_rgb" type="fixed">
    <parent link="depth_camera_link"/>
    <child link="rgb_camera_link"/>
    <origin rpy="0. 0. 0." xyz="0. 0. 0."/>
  </joint>
  <joint name="base_kinect_joint" type="fixed">
    <parent link="panda_link0"/>
    <child link="kinect_base"/>
    <origin rpy="0.05807847712353446 0.32117765639844126 -3.083483219329227"
            xyz="1.5751255618818367 0.09100056881597385 0.6486937792679903"/>
  </joint>
  <gazebo reference="kinect_visor">
    <sensor type="depth" name="kinect">
      <camera>
        <horizontal_fov>1.2108</horizontal_fov>
        <image>
          <width>1280</width>
          <height>720</height>
        </image>
        <clip>
          <near>0.01</near>
          <far>100</far>
        </clip>
      </camera>
    </sensor>
  </gazebo>

  <material name="aluminum">
    <color rgba="0.5 0.5 0.5 1"/>
  </material>
  <material name="plastic">
    <color rgba="0.1 0.1 0.1 1"/>
  </material>
  <joint name="realsense_link_joint" type="fixed">
    <origin rpy="-0.3482024436481727 -1.563844186046295 2.713723394878007"
            xyz="0.022938977261424617 -0.04925718426893216 0.05882856047291339"/>
    <parent link="panda_link8"/>
    <child link="realsense_link"/>
  </joint>
  <link name="realsense_link">
    <visual>
      <!-- the mesh origin is at front plate in between the two infrared realsense axes -->
      <origin rpy="1.5707963267948966 0 1.5707963267948966" xyz="0.0043 -0.0175 0"/>
      <geometry>
        <mesh filename="package://realsense2_description/meshes/d435.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 -0.0175 0"/>
      <geometry>
        <box size="0.02505 0.09 0.025"/>
      </geometry>
    </collision>
    <inertial>
      <!-- The following are not reliable values, and should not be used for modeling -->
      <mass value="0.072"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.003881243" ixy="0.0" ixz="0.0" iyy="0.000498940" iyz="0.0" izz="0.003879257"/>
    </inertial>
  </link>
  <!-- realsense depth joints and links -->
  <joint name="realsense_depth_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="realsense_link"/>
    <child link="realsense_depth_frame"/>
  </joint>
  <link name="realsense_depth_frame"/>
  <joint name="realsense_depth_optical_joint" type="fixed">
    <origin rpy="-1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0"/>
    <parent link="realsense_depth_frame"/>
    <child link="realsense_depth_optical_frame"/>
  </joint>
  <link name="realsense_depth_optical_frame"/>
  <!-- realsense color joints and links -->
  <joint name="realsense_color_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="realsense_link"/>
    <child link="realsense_color_frame"/>
  </joint>
  <link name="realsense_color_frame"/>
  <joint name="realsense_color_optical_joint" type="fixed">
    <origin rpy="-1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0"/>
    <parent link="realsense_color_frame"/>
    <child link="realsense_color_optical_frame"/>
  </joint>
  <link name="realsense_color_optical_frame"/>
  <gazebo reference="realsense_depth_frame">
    <sensor type="depth" name="realsense">
      <camera>
        <horizontal_fov>1.2108</horizontal_fov>
        <image>
          <width>1280</width>
          <height>720</height>
        </image>
        <clip>
          <near>0.01</near>
          <far>100</far>
        </clip>
      </camera>
    </sensor>
  </gazebo>
</robot>
